actionpack/lib/action_dispatch/journey/router.rb

```diff
@@ -45,7 +45,13 @@ def serve(req)
             tmp_params[key] = val.force_encoding(::Encoding::UTF_8)
           }
 
-          req.path_parameters = tmp_params
+          # Validate path, querystring and request params
+          req.path_parameters = encode_parameters_and_validate(req, tmp_params, tmp_params, "path")
+          if req.get?
+            req.query_parameters = encode_parameters_and_validate(req, req.query_parameters, tmp_params, "query")
+          elsif req.post?
+            req.request_parameters = encode_parameters_and_validate(req, req.request_parameters, tmp_params, "request")
+          end
```

Comment:
Aha! I think what's happening is that there can be multiple matching routes here and we might make this check multiple times.

In my case I believe this is happening because of a mounted engine at the root

Reply:
Maybe we could solve this by moving the logic further down into `ActionDispatch::Routing::RouteSet::Dispatcher`?